### PR TITLE
Implement pgrange union 

### PR DESCRIPTION
**Pull request by vijairamg**
Implements `anyrange + anyrange -> anyrange`, `anyrange * anyrange -> anyrange` and `anyrange - anyrange -> anyrange` part of #4092

### CODE REVIEW

File: diesel/src/pg/expression/operators.rs [hunk 1]
```diff
@@ -118,3 +116,49 @@ where
...
+    }
+}
+
+impl<L, R, ST> crate::expression::Expression for UnionRange<L, R>
+where
+    L: crate::expression::Expression<SqlType = Range<ST>>,
+    R: crate::expression::Expression<SqlType = Range<ST>>,
+    ST: SqlType + TypedExpressionType,
+{
+    type SqlType = ST;
```
Comment (weiznich) [inline diesel/src/pg/expression/operators.rs:142]: That's the problematic line that's causing the error.

You specify here that `range_a.union_range(range_b)` return the inner type of the range instead. You want `type SqlType = Range<ST>` instead, which just says that the operator returns a range as well.

That written: You should likely use the `infix_operator!()` macro instead to generate all this code for you. It should be much easier that way.

Comment (vijairamg) [reply diesel/src/pg/expression/operators.rs]: Thanks for the pointer.
My bad! Should have noticed the `__diesel_infix_operator!`.

**Comment by vijairamg** [comment]
Let me know if any of the names need to be changed.

File: diesel/src/pg/expression/expression_methods.rs [hunk 2]
```diff
@@ -1074,6 +1074,150 @@ pub trait PgRangeExpressionMethods: Expression + Sized {
...
+    where
+        Self::SqlType: SqlType,
+        T: AsExpression<Self::SqlType>,
+    {
+        Grouped(DifferenceRange::new(self, other.as_expression()))
+    }
+
+    /// Creates a PostgreSQL ` * ` expression.
+    ///
+    /// This operator takes two ranges and returns the difference.
```
Comment (weiznich) [inline diesel/src/pg/expression/expression_methods.rs:1176]:
```suggestion
    /// This operator takes two ranges and returns the intersection.
```

File: diesel/src/pg/expression/expression_methods.rs [hunk 3]
```diff
@@ -1074,6 +1074,150 @@ pub trait PgRangeExpressionMethods: Expression + Sized {
...
+    where
+        Self::SqlType: SqlType,
+        T: AsExpression<Self::SqlType>,
+    {
+        Grouped(UnionsRange::new(self, other.as_expression()))
+    }
+
+    /// Creates a PostgreSQL ` - ` expression.
+    ///
+    /// This operator takes two ranges and returns the difference.
```
Comment (weiznich) [inline diesel/src/pg/expression/expression_methods.rs:1128]:
```suggestion
    /// This operator takes two ranges and returns the difference.
    /// 
    /// The second range must not be contained in the first in such a way that the 
    /// difference would not be a single range.
```

**Review by weiznich — APPROVED** [review]
Looks good beside the two minor documentation fixes. Thanks for working on this :+1:

**Pull request merged**